torch_xla/experimental/apply_layers.py

```diff
@@ -0,0 +1,126 @@
+from typing import Iterable, Mapping, Sequence, Dict, Tuple
+
+import torch
+import torch.nn as nn
+from torch.utils._pytree import tree_map
+from functorch.compile import default_partition
+
+import torch_xla
+from torch_xla.experimental.scan import scan
+
+
+def apply_layers(layers: Iterable[torch.nn.Module],
+                 input_data,
+                 partition_fn=default_partition):
+  """Applies each layer in `layers` to `input_data` sequentially.
+
+  `input_data` is provided as input to the first layer in `layers`. The output of one
+  layer is provided as input to next layer.
+
+  All modules in `layers` must have the same structure, and they must perform the same
+  calculations given the same model parameters and inputs. In practice, this means you
+  cannot use different dropout probabilities, parameter shapes, activation functions etc.,
+  across the `layers`.
+
+  Under these conditions, this function is equivalent to
+
+    sequential = torch.nn.Sequential(*layers)
+    sequential(input_data)
+
+  This function can be faster to compile since it reuses the XLA computation of the
+  first layer to perform the computation of all other layers.
+
+  ## partition_fn
+
+  Since this function uses AOTAutograd to trace `fn`, you may override what computation
+  happen in the forward and backward passes by specifying different partition functions.
+  `default_partition` implies no activation checkpointing. You may specify
+  `functorch.compile.min_cut_rematerialization_partition` to use min-cut based activation
+  checkpointing. You may also write your own partitioner to insert any custom logic such
+  as host offloading of activations.
+  """
```

Comment:
I understand `partition_fn` is a more accurate name, but from a UX perspective you might want to just make it a book to enable gradient checkpointing or not.

Reply:
Here's my plan:

- Need to keep `partition_fn` around so I could also inject offloading logic as I partition the graph.
- Later, integrate with dynamo to get rid of this extra function altogether. Dynamo can decompose `torch.utils.checkpoint` into proper fw/bw graphs.
- When scan moves out of experimental, this arg will go away.